source/opt/constant_value.h

```diff
@@ -0,0 +1,304 @@
+  std::vector<uint32_t> words_;
+};
+
+// Integer type constant.
+class IntConst : public ScalarConst {
+ public:
+  IntConst(const analysis::Integer* ty, const std::vector<uint32_t>& w)
+      : ScalarConst(ty, w) {}
+  IntConst(const analysis::Integer* ty, std::vector<uint32_t>&& w)
+      : ScalarConst(ty, std::move(w)) {}
+
+  IntConst* AsIntConst() override { return this; }
+  const IntConst* AsIntConst() const override { return this; }
+
+  std::unique_ptr<IntConst> CopyIntConst() const {
+    return MakeUnique<IntConst>(type_->AsInteger(), words_);
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyIntConst().release());
+  }
+};
+
+// Float type constant.
+class FloatConst : public ScalarConst {
+ public:
+  FloatConst(const analysis::Float* ty, const std::vector<uint32_t>& w)
+      : ScalarConst(ty, w) {}
+  FloatConst(const analysis::Float* ty, std::vector<uint32_t>&& w)
+      : ScalarConst(ty, std::move(w)) {}
+
+  FloatConst* AsFloatConst() override { return this; }
+  const FloatConst* AsFloatConst() const override { return this; }
+
+  std::unique_ptr<FloatConst> CopyFloatConst() const {
+    return MakeUnique<FloatConst>(type_->AsFloat(), words_);
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyFloatConst().release());
+  }
+};
+
+// Bool type constant.
+class BoolConst : public ScalarConst {
+ public:
+  BoolConst(const analysis::Bool* ty, bool v)
+      : ScalarConst(ty, {static_cast<uint32_t>(v)}), val_(v) {}
+
+  BoolConst* AsBoolConst() override { return this; }
+  const BoolConst* AsBoolConst() const override { return this; }
+
+  std::unique_ptr<BoolConst> CopyBoolConst() const {
+    return MakeUnique<BoolConst>(type_->AsBool(), val_);
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyBoolConst().release());
+  }
+
+  bool val() const { return val_; }
+
+ private:
+  bool val_;
+};
+
+// Abstract class for composite constants.
+class CompositeConst : public Const {
+ public:
+  CompositeConst() = delete;
+  CompositeConst* AsCompositeConst() override { return this; }
+  const CompositeConst* AsCompositeConst() const override { return this; }
+
+  virtual uint32_t NumComponents() const {
+    return static_cast<uint32_t>(components_.size());
+  }
+  virtual const Const* GetComponent(uint32_t index) const {
+    assert(static_cast<size_t>(index) < components_.size() &&
+           "GetComponent() index out of bound.");
+    return components_[index];
+  }
+
+ protected:
+  CompositeConst(const analysis::Type* ty) : Const(ty), components_() {}
+  CompositeConst(const analysis::Type* ty,
+                 const std::vector<const Const*>& components)
+      : Const(ty), components_(components) {}
+  CompositeConst(const analysis::Type* ty,
+                 std::vector<const Const*>&& components)
+      : Const(ty), components_(std::move(components)) {}
+  std::vector<const Const*> components_;
```

Comment:
It's a little worrying that components_ is not private. It very tightly couples the state of this object with its subclasses.  But perhaps there isn't much harm (for now).  Alternatives would require more code and indirection.